src/TurbulenceClosures/turbulence_closure_implementations/CATKEVerticalDiffusivities/turbulent_kinetic_energy_equation.jl

```diff
@@ -24,35 +73,61 @@ end
     N² = ℑzᵃᵃᶜ(i, j, k, grid, ∂z_b, buoyancy, tracers)
     return - κᶻ * N²
 end
+=#
 
 const VITD = VerticallyImplicitTimeDiscretization
 
-@inline dissipation(i, j, k, grid, closure::FlavorOfCATKE{<:VITD}, args...) = zero(eltype(grid))
+@inline function buoyancy_flux(i, j, k, grid, closure::FlavorOfCATKE{<:VITD}, velocities, tracers, buoyancy, diffusivities)
+    wb = ℑzᵃᵃᶜ(i, j, k, grid, buoyancy_fluxᶜᶜᶠ, tracers, buoyancy, diffusivities)
+    
+    #κᶻ = ℑzᵃᵃᶜ(i, j, k, grid, diffusivities.Kᶜ)
+    #N² = ℑzᵃᵃᶜ(i, j, k, grid, ∂z_b, buoyancy, tracers)
+    #wb = - κᶻ * N²
+    
+    eⁱʲᵏ = @inbounds tracers.e[i, j, k]
+    wb_eⁱʲᵏ = wb / eⁱʲᵏ
+
+    # "Patankar trick" for buoyancy production (cf Patankar 1980 or Burchard et al. 2003)
+    # If buoyancy flux is a _sink_ of TKE, we treat it implicitly.
+    return ifelse(sign(wb) * sign(eⁱʲᵏ) < 0, zero(grid), wb)
+end
+
+@inline dissipation(i, j, k, grid, closure::FlavorOfCATKE{<:VITD}, args...) = zero(grid)
 
-@inline function implicit_dissipation_coefficient(i, j, k, grid, closure::FlavorOfCATKE{<:VITD}, velocities, tracers, buoyancy, clock, tracer_bcs)
+@inline function implicit_dissipation_coefficient(i, j, k, grid, closure::FlavorOfCATKE{<:VITD},
+                                                  velocities, tracers, buoyancy, clock, tracer_bcs)
     e = tracers.e
     FT = eltype(grid)
-    @inbounds e⁺ = abs(e[i, j, k])
 
-    ℓ = ℑzᵃᵃᶜ(i, j, k, grid, TKE_mixing_lengthᶜᶜᶠ, closure, velocities, tracers, buoyancy, clock, tracer_bcs)
-    Cᴰ = closure.Cᴰ
+    # Start with tracer mixing length
+    ℓ = ℑzᵃᵃᶜ(i, j, k, grid, tracer_mixing_lengthᶜᶜᶠ, closure, velocities, tracers, buoyancy, clock, tracer_bcs)
+
+    # Ri-dependent dissipation coefficient
+    Cᴰ⁻ = closure.turbulent_kinetic_energy_equation.Cᴰ⁻
+    Cᴰ⁺ = closure.turbulent_kinetic_energy_equation.Cᴰ⁺
+    Riᶜ = closure.turbulent_kinetic_energy_equation.CᴰRiᶜ
+    Riʷ = closure.turbulent_kinetic_energy_equation.CᴰRiʷ
+    Ri = Riᶜᶜᶜ(i, j, k, grid, velocities, tracers, buoyancy)
+    Cᴰ = scale(Ri, Cᴰ⁻, Cᴰ⁺, Riᶜ, Riʷ)
+
+    eᵢ = @inbounds e[i, j, k]
 
     # Note:
-    #   * because   ∂t e + ⋯ = ⋯ + L e = ⋯ - ϵ,
+    #   Because   ∂t e + ⋯ = ⋯ + L e = ⋯ - ϵ,
     #
-    #   * then      L e = - ϵ
-    #                   = - Cᴰ e³² / ℓ
+    #   then      L e = - ϵ
+    #                 = - Cᴰ e³² / ℓ
     #
-    #   * and thus    L = - Cᴰ √e / ℓ
-    
-    return - Cᴰ * sqrt(e⁺) / ℓ
+    #   and thus    L = - Cᴰ √e / ℓ .
+
+    return - Cᴰ * sqrt(abs(eᵢ)) / ℓ
```

Comment:
why `abs`? `eᵢ` could be negative? is it just for small numerical errors?

Reply:
Yes, due to time discretization errors `e` can be negative (unfortunately).